trailmet/algorithms/quantize/reconstruct.py

```diff
@@ -0,0 +1,355 @@
+
+        b = self.temp_decay(self.count)
+        if self.count < self.loss_start or self.round_loss == 'none':
+            b = round_loss = 0
+        elif self.round_loss == 'relaxation':
+            round_loss = 0
+            round_vals = self.layer.weight_quantizer.get_soft_targets()
+            round_loss += self.weight * (1 - ((round_vals - .5).abs() * 2).pow(b)).sum()
+        else:
+            raise NotImplementedError
+
+        total_loss = rec_loss + round_loss
+        if self.count % 1000 == 0:
+            print('Total loss:\t{:.3f} (rec:{:.3f}, round:{:.3f})\tb={:.2f}\tcount={}'.format(
+                  float(total_loss), float(rec_loss), float(round_loss), b, self.count))
+        return total_loss
+
+
+def layer_reconstruction(model: QuantModel, layer: QuantModule, cali_data: torch.Tensor,
+                         batch_size: int = 32, iters: int = 20000, weight: float = 0.001, opt_mode: str = 'mse',
+                         asym: bool = False, include_act_func: bool = True, b_range: tuple = (20, 2),
+                         warmup: float = 0.0, act_quant: bool = False, lr: float = 4e-5, p: float = 2.0,
+                         multi_gpu: bool = False):
+    """
+    Block reconstruction to optimize the output from each layer.
+    :param model: QuantModel
+    :param layer: QuantModule that needs to be optimized
+    :param cali_data: data for calibration, typically 1024 training images, as described in AdaRound
+    :param batch_size: mini-batch size for reconstruction
+    :param iters: optimization iterations for reconstruction,
+    :param weight: the weight of rounding regularization term
+    :param opt_mode: optimization mode
+    :param asym: asymmetric optimization designed in AdaRound, use quant input to reconstruct fp output
+    :param include_act_func: optimize the output after activation function
+    :param b_range: temperature range
+    :param warmup: proportion of iterations that no scheduling for temperature
+    :param act_quant: use activation quantization or not.
+    :param lr: learning rate for act delta learning
+    :param p: L_p norm minimization
+    :param multi_gpu: use multi-GPU or not, if enabled, we should sync the gradients
+    """
+
+    model.set_quant_state(False, False)
+    layer.set_quant_state(True, act_quant)
+    round_mode = 'learned_hard_sigmoid'
+
+    if not include_act_func:
+        org_act_func = layer.activation_function
+        layer.activation_function = StraightThrough()
+
+    if not act_quant:
+        # Replace weight quantizer to AdaRoundQuantizer
+        layer.weight_quantizer = AdaRoundQuantizer(uaq=layer.weight_quantizer, round_mode=round_mode,
+                                                weight_tensor=layer.org_weight.data)
+        layer.weight_quantizer.soft_targets = True
+
+        # Set up optimizer
+        opt_params = [layer.weight_quantizer.alpha]
+        optimizer = torch.optim.Adam(opt_params)
```

Comment:
consider adding optimizer to kwargs

Reply:
done in 33615aff540d4b3883842eca67b1a56538f591a1